File: dana.core/dana/tests/event.py
#!/usr/bin/env python
#
# Copyright (c) 2006-2007 Nicolas Rougier.
# All rights reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
# 
# $Id: unit.py 278 2007-08-17 06:23:47Z rougier $

import unittest
import dana.core as core


class EventTests (unittest.TestCase):
    def setUp (self):
        self.model = core.Model()
        self.network = core.Network()
        self.map = core.Map ( (10,10) )
        self.layer = core.Layer()
        self.model.append (self.network)
        self.network.append (self.map)
        self.map.append (self.layer)
        self.layer.fill (core.Unit)
        
   
    def testEventDP (self):
        """ EventDP triggered by layer """
        class Observer(core.Observer):
            count = 0
            def notify (self, event):
                self.count += 1
        obs = Observer()
        core.EventDP.attach (obs)
        self.layer.compute_dp()
        self.assertEqual (obs.count, 100)

    def testEventDW (self):
        """ EventDW triggered by layer """
        class Observer(core.Observer):
            count = 0
            def notify (self, event):
                self.count += 1
        obs = Observer()
        core.EventDW.attach (obs)
        self.layer.compute_dw()
        self.assertEqual (obs.count, 100)


# Test suite
suite = unittest.TestLoader().loadTestsFromTestCase(EventTests)

if __name__ == "__main__":
    unittest.main()

File: dana.core/dana/core/event.cc
/*
  DANA - Distributed Asynchronous Numerical Adaptive computing library
  Copyright (c) 2006,2007,2008 Nicolas P. Rougier

  This file is part of DANA.

  DANA is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your
  option) any later version.

  DANA is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
  for more details.

  You should have received a copy of the GNU General Public License
  along with DANA. If not, see <http://www.gnu.org/licenses/>.
*/

#include <algorithm>
#include "event.h"
#include "observer.h"

using namespace dana::core;


// _______________________________________________________________________export
void
Event::python_export (void)
{
    using namespace boost::python;
    register_ptr_to_python< boost::shared_ptr<Event> >();
    register_ptr_to_python< boost::shared_ptr<EventDP> >();
    register_ptr_to_python< boost::shared_ptr<EventDW> >();
    register_ptr_to_python< boost::shared_ptr<EventEvaluate> >();
 
    class_<Event, bases <Object> >(
        "Event",
        "______________________________________________________________________\n"
        "                                                                      \n"
        "______________________________________________________________________\n",
        init<>("__init__()"))
        ;

    class_<EventDP, bases <Event> >(
        "EventDP",
        "______________________________________________________________________\n"
        "                                                                      \n"
        "______________________________________________________________________\n",
        init<>("__init__()"))
        ;


    class_<EventDW, bases <Event> >(
        "EventDW",
        "______________________________________________________________________\n"
        "                                                                      \n"
        "______________________________________________________________________\n",
        init<>("__init__()"))
        ;

    class_<EventEvaluate, bases <Event> >(
        "EventEvaluate",
        "______________________________________________________________________\n"
        "                                                                      \n"
        "______________________________________________________________________\n",
        init<>("__init__()"))
        ;

}

